BleScanException: Scan failed because feature unsupported on api 23+ due to CALLBACK_TYPE_FIRST_MATCH | CALLBACK_TYPE_MATCH_LOST

### Summary
Starting a scan with `CALLBACK_TYPE_FIRST_MATCH | CALLBACK_TYPE_MATCH_LOST` errors with `SCAN_FAILED_FEATURE_UNSUPPORTED` or `SCAN_FAILED_OUT_OF_HARDWARE_RESOURCES`

Testing on API < 21 works without problems, and while looking at the code in `ScanSetupBuilderImplApi18` it is working because the behavior is being emulated.

`ScanSetupBuilderImplApi21` also emulates said behavior, although I cannot test it.

`ScanSetupBuilderImplApi23` does NOT emulate callbacks, on the basis that they are natively available.

Thing is all my available Android API 23+ test devices are old or just plain cheap and their Bluetooth chips likely do not support said behavior, which would make sense given the thrown errors.

I think `ScanSetupBuilderImplApi23` should retry on error with the emulated behavior, or at least give access to it outside of the library, so it can be implemented without having to copy the same functionality.

Thoughts?

### Library version
`1.9.1`

#### Preconditions
```
ScanSettings.Builder()
                .setCallbackType(ScanSettings.CALLBACK_TYPE_FIRST_MATCH or ScanSettings.CALLBACK_TYPE_MATCH_LOST)
                .build()
```

#### Steps to reproduce actual result

#### Minimum code snippet reproducing the issue
```
client
        .scanBleDevices(
            ScanSettings.Builder()
                .setCallbackType(ScanSettings.CALLBACK_TYPE_FIRST_MATCH or ScanSettings.CALLBACK_TYPE_MATCH_LOST)
                .setScanMode(ScanSettings.SCAN_MODE_BALANCED)
                .build(),
            *filters
        )
        .subscribe()
```
#### Logs from the application running with setting 
```
2019-04-12 13:05:19.540 10603-10636/[..] D/RxBle#ClientOperationQueue: QUEUED   ScanOperationApi21(158477343)
2019-04-12 13:05:19.541 10603-10627/[..] D/RxBle#ClientOperationQueue: STARTED  ScanOperationApi21(158477343)
2019-04-12 13:05:19.542 10603-10627/[..] I/RxBle#ClientOperationQueue: RUNNING  ScanOperationApi21{ANY_MUST_MATCH -> nativeFilters=[BluetoothLeScanFilter [mDeviceName=null, MAC=null, mUuid=null, mUuidMask=null, mServiceDataUuid=null, mServiceData=null, mServiceDataMask=null, mManufacturerId=-1, mManufacturerData=null, mManufacturerDataMask=null]]}
2019-04-12 13:05:19.549 10603-10637/[..] I/RxBle#QueueOperation: Scan operation is requested to start.
2019-04-12 13:05:19.549 10603-10637/[..] D/RxBle#ScanOperationApi21: No library side filtering —> debug logs of scanned devices disabled
2019-04-12 13:05:19.557 10603-10627/[..] D/RxBle#ClientOperationQueue: FINISHED ScanOperationApi21(158477343) in 16 ms
2019-04-12 13:05:19.598 10603-10637/[..] I/RxBle#CancellableDisposable: Scan operation is requested to stop.
2019-04-12 13:05:19.601 10603-10637/[..] D/BluetoothLeScanner: could not find callback wrapper
2019-04-12 13:05:19.601 10603-10603/[..] W/System.err: io.reactivex.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | com.polidea.rxandroidble2.exceptions.BleScanException: Scan failed because feature unsupported (code 8)
2019-04-12 13:05:19.602 10603-10603/[..] W/System.err:     at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
2019-04-12 13:05:19.602 10603-10603/[..] W/System.err:     at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
2019-04-12 13:05:19.602 10603-10603/[..] W/System.err:     at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
2019-04-12 13:05:19.602 10603-10603/[..] W/System.err:     at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:281)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:172)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:124)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at android.os.Handler.handleCallback(Handler.java:751)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:95)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at android.os.Looper.loop(Looper.java:154)
2019-04-12 13:05:19.603 10603-10603/[..] W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6126)
2019-04-12 13:05:19.604 10603-10603/[..] W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2019-04-12 13:05:19.604 10603-10603/[..] W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
2019-04-12 13:05:19.604 10603-10603/[..] W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
2019-04-12 13:05:19.604 10603-10603/[..] W/System.err: Caused by: com.polidea.rxandroidble2.exceptions.BleScanException: Scan failed because feature unsupported (code 8)
2019-04-12 13:05:19.605 10603-10603/[..] W/System.err:     at com.polidea.rxandroidble2.internal.operations.ScanOperationApi21$1.onScanFailed(ScanOperationApi21.java:92)
2019-04-12 13:05:19.605 10603-10603/[..] W/System.err:     at android.bluetooth.le.BluetoothLeScanner$1.run(BluetoothLeScanner.java:461)
2019-04-12 13:05:19.605 10603-10603/[..] W/System.err: 	... 7 more
```
#### Actual result
`com.polidea.rxandroidble2.exceptions.BleScanException: Scan failed because feature unsupported (code 8)`

#### Expected result
Library emulates behavior (on error) in API 23 just like pre API 23

--- a/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ExternalScanSettingsExtension.java
+++ b/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ExternalScanSettingsExtension.java
@@ -7,6 +7,14 @@ public interface ExternalScanSettingsExtension<R extends ExternalScanSettingsExt
 
     boolean shouldCheckLocationProviderState();
 
+    // [DS 18.09.2019] Introduced to be sure that new ScanSettings properties will not break workaround introduced in
+    // ScanSettingsBuilderImplApi21
+    /**
+     * Copies the current ScanSettings with changed callback type.
+     *
+     * @param callbackType callback type of the copied object
+     * @return new ScanSettings object with copied properties and new callback type
+     */
     R copyWithCallbackType(int callbackType);
 
     interface Builder<T extends Builder<T>> {

--- a/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ExternalScanSettingsExtension.java
+++ b/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ExternalScanSettingsExtension.java
@@ -3,9 +3,11 @@ package com.polidea.rxandroidble2.internal.scan;
 /**
  * An interface that describes what library extensions should be added to {@link com.polidea.rxandroidble2.scan.ScanSettings}
  */
-public interface ExternalScanSettingsExtension {
+public interface ExternalScanSettingsExtension<R extends ExternalScanSettingsExtension<R>> {
 
     boolean shouldCheckLocationProviderState();
+
+    R copyWithCallbackType(int callbackType);
 
     interface Builder<T extends Builder<T>> {
 

--- a/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ScanSetupBuilderImplApi23.java
+++ b/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ScanSetupBuilderImplApi23.java
@@ -39,22 +39,23 @@ public class ScanSetupBuilderImplApi23 implements ScanSetupBuilder {
     @RequiresApi(21 /* Build.VERSION_CODES.LOLLIPOP */)
     @Override
     public ScanSetup build(ScanSettings scanSettings, ScanFilter... scanFilters) {
-        // for now assuming that on Android 6.0+ there are no other problems than commented below
-        boolean scanFiltersEmpty = true;
-        for (ScanFilter scanFilter : scanFilters) {
-            scanFiltersEmpty &= scanFilter.isAllFieldsEmpty();
-        }
+        boolean areFiltersSpecified = areFiltersSpecified(scanFilters);
+        boolean isFilteringCallbackType = scanSettings.getCallbackType() != ScanSettings.CALLBACK_TYPE_ALL_MATCHES;
 
         ObservableTransformer<RxBleInternalScanResult, RxBleInternalScanResult> resultTransformer = ObservableUtil.identityTransformer();
         ScanSettings resultScanSettings = scanSettings;
-        if (scanSettings.getCallbackType() != ScanSettings.CALLBACK_TYPE_ALL_MATCHES && scanFiltersEmpty) {
-            // native matching does not work with no filters specified - see https://issuetracker.google.com/issues/37127640
-            // so we will use a callback type that will work and emulate the desired behaviour
-            RxBleLog.d("ScanSettings.callbackType != CALLBACK_TYPE_ALL_MATCHES but no filters specified. "
+
+        // native matching (when a device is first seen or no longer seen) does not work with no filters specified —
+        // see https://issuetracker.google.com/issues/37127640
+        // so we will use a callback type that will work and emulate the desired behaviour
+        boolean shouldEmulateCallbackType = isFilteringCallbackType && !areFiltersSpecified;
+        if (shouldEmulateCallbackType) {
+            RxBleLog.d("ScanSettings.callbackType != CALLBACK_TYPE_ALL_MATCHES but no (or only empty) filters are specified. "
                 + "Falling back to callbackType emulation.");
             resultTransformer = scanSettingsEmulator.emulateCallbackType(scanSettings.getCallbackType());
             resultScanSettings = scanSettings.copyWithCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES);
         }
+
         return new ScanSetup(
                 new ScanOperationApi21(
                         rxBleAdapterWrapper,
@@ -65,5 +66,13 @@ public class ScanSetupBuilderImplApi23 implements ScanSetupBuilder {
                         scanFilters),
                 resultTransformer
         );
+    }
+
+    private boolean areFiltersSpecified(ScanFilter[] scanFilters) {
+        boolean scanFiltersEmpty = true;
+        for (ScanFilter scanFilter : scanFilters) {
+            scanFiltersEmpty &= scanFilter.isAllFieldsEmpty();
+        }
+        return !scanFiltersEmpty;
     }
 }

--- a/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ScanSetupBuilderImplApi23.java
+++ b/rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/scan/ScanSetupBuilderImplApi23.java
@@ -4,14 +4,15 @@ package com.polidea.rxandroidble2.internal.scan;
 import androidx.annotation.RequiresApi;
 import androidx.annotation.RestrictTo;
 
+import com.polidea.rxandroidble2.internal.RxBleLog;
 import com.polidea.rxandroidble2.internal.operations.ScanOperationApi21;
+import com.polidea.rxandroidble2.internal.util.ObservableUtil;
 import com.polidea.rxandroidble2.internal.util.RxBleAdapterWrapper;
 import com.polidea.rxandroidble2.scan.ScanFilter;
 import com.polidea.rxandroidble2.scan.ScanSettings;
 
 import bleshadow.javax.inject.Inject;
 
-import io.reactivex.Observable;
 import io.reactivex.ObservableTransformer;
 
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
@@ -19,44 +20,50 @@ public class ScanSetupBuilderImplApi23 implements ScanSetupBuilder {
 
     private final RxBleAdapterWrapper rxBleAdapterWrapper;
     private final InternalScanResultCreator internalScanResultCreator;
+    private final ScanSettingsEmulator scanSettingsEmulator;
     private final AndroidScanObjectsConverter androidScanObjectsConverter;
 
     @Inject
     ScanSetupBuilderImplApi23(
             RxBleAdapterWrapper rxBleAdapterWrapper,
             InternalScanResultCreator internalScanResultCreator,
+            ScanSettingsEmulator scanSettingsEmulator,
             AndroidScanObjectsConverter androidScanObjectsConverter
     ) {
         this.rxBleAdapterWrapper = rxBleAdapterWrapper;
         this.internalScanResultCreator = internalScanResultCreator;
+        this.scanSettingsEmulator = scanSettingsEmulator;
         this.androidScanObjectsConverter = androidScanObjectsConverter;
     }
 
     @RequiresApi(21 /* Build.VERSION_CODES.LOLLIPOP */)
     @Override
     public ScanSetup build(ScanSettings scanSettings, ScanFilter... scanFilters) {
-        // for now assuming that on Android 6.0+ there are no problems
+        // for now assuming that on Android 6.0+ there are no other problems than commented below
+        boolean scanFiltersEmpty = true;
+        for (ScanFilter scanFilter : scanFilters) {
+            scanFiltersEmpty &= scanFilter.isAllFieldsEmpty();
+        }
 
-        if (scanSettings.getCallbackType() != ScanSettings.CALLBACK_TYPE_ALL_MATCHES && scanFilters.length == 0) {
+        ObservableTransformer<RxBleInternalScanResult, RxBleInternalScanResult> resultTransformer = ObservableUtil.identityTransformer();
+        ScanSettings resultScanSettings = scanSettings;
+        if (scanSettings.getCallbackType() != ScanSettings.CALLBACK_TYPE_ALL_MATCHES && scanFiltersEmpty) {
             // native matching does not work with no filters specified - see https://issuetracker.google.com/issues/37127640
-            scanFilters = new ScanFilter[] {
-                    ScanFilter.empty()
-            };
+            // so we will use a callback type that will work and emulate the desired behaviour
+            RxBleLog.d("ScanSettings.callbackType != CALLBACK_TYPE_ALL_MATCHES but no filters specified. "
+                + "Falling back to callbackType emulation.");
+            resultTransformer = scanSettingsEmulator.emulateCallbackType(scanSettings.getCallbackType());
+            resultScanSettings = scanSettings.copyWithCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES);
         }
         return new ScanSetup(
                 new ScanOperationApi21(
                         rxBleAdapterWrapper,
                         internalScanResultCreator,
                         androidScanObjectsConverter,
-                        scanSettings,
+                        resultScanSettings,
                         new EmulatedScanFilterMatcher(),
                         scanFilters),
-                new ObservableTransformer<RxBleInternalScanResult, RxBleInternalScanResult>() {
-                    @Override
-                    public Observable<RxBleInternalScanResult> apply(Observable<RxBleInternalScanResult> observable) {
-                        return observable;
-                    }
-                }
+                resultTransformer
         );
     }
 }

--- a/rxandroidble/src/main/java/com/polidea/rxandroidble2/scan/ScanSettings.java
+++ b/rxandroidble/src/main/java/com/polidea/rxandroidble2/scan/ScanSettings.java
@@ -19,7 +19,7 @@ import java.lang.annotation.RetentionPolicy;
  * https://code.google.com/p/android/issues/detail?id=178614
  * https://code.google.com/p/android/issues/detail?id=228428
  */
-public class ScanSettings implements Parcelable, ExternalScanSettingsExtension {
+public class ScanSettings implements Parcelable, ExternalScanSettingsExtension<ScanSettings> {
 
     @IntDef({SCAN_MODE_OPPORTUNISTIC, SCAN_MODE_LOW_POWER, SCAN_MODE_BALANCED, SCAN_MODE_LOW_LATENCY})
     @Retention(RetentionPolicy.SOURCE)
@@ -222,6 +222,18 @@ public class ScanSettings implements Parcelable, ExternalScanSettingsExtension {
             return new ScanSettings(in);
         }
     };
+
+    @Override
+    public ScanSettings copyWithCallbackType(@CallbackType int callbackType) {
+        return new ScanSettings(
+                this.mScanMode,
+                callbackType,
+                this.mReportDelayMillis,
+                this.mMatchMode,
+                this.mNumOfMatchesPerFilter,
+                this.mShouldCheckLocationProviderState
+        );
+    }
 
     /**
      * Builder for {@link ScanSettings}.

